File: solutions/_101_symmetric_tree/Solution.java
package _101_symmetric_tree;

class Solution {

  public boolean isSymmetric(TreeNode root) {
    return isMirror(root, root);
  }

  private boolean isMirror(TreeNode left, TreeNode right) {
    if (left == null && right == null) {
      return true;
    }
    if (left == null || right == null) {
      return false;
    }
    return (
      (left.val == right.val) &&
      isMirror(left.right, right.left) &&
      isMirror(left.left, right.right)
    );
  }
}
/*
https://leetcode.com/problems/symmetric-tree/
*/

File: solutions/_101_symmetric_tree/solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        return self.is_mirror(root.left, root.right)

    def is_mirror(self, left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
        if not left and not right:
            return True
        if not left or not right:
            return False
        return (
            left.val == right.val
            and self.is_mirror(left.left, right.right)
            and self.is_mirror(left.right, right.left)
        )


# https://leetcode.com/problems/symmetric-tree/
